packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts

```diff
@@ -0,0 +1,530 @@
+    for (const rootHex of this.epochIndex.get(cp.epoch)?.values() || []) {
+      const cpKey = toCheckpointKey({epoch: cp.epoch, rootHex});
+      const stateOrPersistentKey = this.cache.get(cpKey);
+      if (stateOrPersistentKey !== undefined && !isPersistentKey(stateOrPersistentKey)) {
+        // this is a state
+        foundState = true;
+        break;
+      }
+    }
+    if (!foundState) {
+      this.inMemoryEpochs.delete(cp.epoch);
+    }
+    const epochKey = toHexString(cp.root);
+    const value = this.epochIndex.get(cp.epoch);
+    if (value) {
+      value.delete(epochKey);
+      if (value.size === 0) {
+        this.epochIndex.delete(cp.epoch);
+      }
+    }
+  }
+
+  /**
+   * Delete all items of an epoch from disk and memory
+   */
+  async deleteAllEpochItems(epoch: Epoch): Promise<void> {
+    for (const rootHex of this.epochIndex.get(epoch) || []) {
+      const key = toCheckpointKey({rootHex, epoch});
+      const stateOrPersistentKey = this.cache.get(key);
+      if (stateOrPersistentKey !== undefined && isPersistentKey(stateOrPersistentKey)) {
+        await this.persistentApis.remove(stateOrPersistentKey);
+        this.metrics?.stateRemoveCount.inc({reason: RemovePersistedStateReason.pruneFinalized});
+      }
+      this.cache.delete(key);
+    }
+    this.inMemoryEpochs.delete(epoch);
+    this.epochIndex.delete(epoch);
+  }
+
+  /**
+   * This is slow code because it involves serializing the whole state to disk which takes 600ms to 900ms on Holesky as of Sep 2023
+   * The add() is called after we process 1st block of an epoch, we don't want to pruneFromMemory at that time since it's the hot time
+   * Call this code at the last 1/3 slot of slot 0 of an epoch
+   */
+  async pruneFromMemory(): Promise<number> {
+    let count = 0;
+    while (this.inMemoryEpochs.size > this.maxEpochsInMemory) {
+      let firstEpoch: Epoch | undefined;
+      for (const epoch of this.inMemoryEpochs) {
+        firstEpoch = epoch;
+        break;
+      }
+      if (firstEpoch === undefined) {
+        // should not happen
+        throw new Error("No epoch in memory");
+      }
+      // first loop to check if the 1st slot of epoch is a skipped slot or not
+      let firstSlotBlockRoot: string | undefined;
+      for (const rootHex of this.epochIndex.get(firstEpoch) ?? []) {
+        const cpKey = toCheckpointKey({epoch: firstEpoch, rootHex});
+        const stateOrPersistentKey = this.cache.get(cpKey);
+        if (stateOrPersistentKey !== undefined && !isPersistentKey(stateOrPersistentKey)) {
+          // this is a state
+          if (
+            rootHex !== toHexString(getBlockRootAtSlot(stateOrPersistentKey, computeStartSlotAtEpoch(firstEpoch) - 1))
+          ) {
+            firstSlotBlockRoot = rootHex;
+            break;
+          }
+        }
+      }
+
+      // if found firstSlotBlockRoot it means it's Current Root Checkpoint State and we should only persist that checkpoint as it's the state
```

Comment:
TODO: even there is Current Root Checkpoint State, it could be from an orphaned block so maybe we still have to store both checkpoints but could improve by:
- store Previous Root Checkpoint State with state bytes
- store Current Root Checkpoint State with root of Previous Root Checkpoint State and block at slot 0 of epoch, and do a replay when we need it